### PR TITLE
compiler: default to `--gc:orc` for the C target

### PR DESCRIPTION
**Pull request by zerbina**
## Summary

After bootstrapping the compiler itself with `--gc:orc`, this is the
second step towards removing the old runtime.

In order to make to make sure that ORC is stable enough, and that the
standard library works with the new runtime, `--gc:orc` is now implied
for the C target when no memory management strategy is explicitly
selected.

### The New Runtime

Using either `--gc:orc` or `--gc:arc` implies usage of the new runtime,
meaning that the new runtime is now the default. The key differences
compared to the old runtime are:
* usage of `deepcopy` in alive code requires the `--deepcopy` option to
  be enabled
* modifying a string through unsafe facilities (pointers and `cstring`)
  requires preparing the string for mutation first (via
  `prepareMutation`)
* `GC_unref` and `GC_ref` are no longer available for `string` and `seq`
  types
* `repr` version 2 is used

With `--gc:orc` and `--gc:arc`, `shallowCopy` is a no-op.

## Details

* support `prepareMutation` for code running in the VM (it's currently a
  no-op)
* fix a crash with `marshal.loadAny` and empty strings, when using the
  new runtime
* disable tests depending on `nimrtl` -- `nimrtl` doesn't yet work with
  the new runtime
* the IC implementation doesn't support type-bound operators (and thus
  `--gc:arc|orc`) at this time, so change its tests to explicitly use
  `--gc:refc`
* adjust the test suite where necessary

### CODE REVIEW

File: compiler/nim.nim [hunk 1]
```diff
@@ -89,6 +90,18 @@ proc handleCmdLine(cache: IdentCache; conf: ConfigRef): CmdLineHandlingResult =
       conf.errorCounter != 0:
     return
 
+  # if no GC was selected by either the command line or the config files,
```
Comment (saem) [inline compiler/nim.nim]: This needs to be true for suggests as well, I'd put it in `cmdlinehelper` for now.

File: tests/stdlib/metaprogramming/mgenast.nim [hunk 2]
```diff
@@ -31,6 +31,7 @@ macro bindme6UseExpose*(): untyped =
   genAst:
     var tst = "sometext"
     var ss = newStringStream("anothertext")
+    prepareMutation(tst)
```
Comment (saem) [inline tests/stdlib/metaprogramming/mgenast.nim:34]: Promise me that if we do something with copy on write it has to be named `prepareMootation`. 🐄

File: tests/stdlib/types/tpackedsets.nim [hunk 3]
```diff
@@ -235,15 +235,15 @@ block legacyMainModuleTests:
 
       proc bug12366 =
         var
-          x = initPackedSet[A]()
+          a = initPackedSet[A]()
```
Comment (saem) [inline tests/stdlib/types/tpackedsets.nim:238]: No need to change it back, just curious why the rename?

**Review by saem — COMMENTED** [review]
I spotted one thing that needs fixing, selecting the gc if none is set as it impacts what's defined or not in stdlib.

Otherwise this is great, very much looking forward to getting this merged in. Thank you very much for persevering. ♥️🚀

File: compiler/nim.nim (hunk 1 above)
Comment (zerbina) [reply compiler/nim.nim]: Ah, thanks for catching this. I forgot about that part.

File: tests/stdlib/types/tpackedsets.nim (hunk 3 above)
Comment (zerbina) [reply tests/stdlib/types/tpackedsets.nim:238]: It's the symbol re-definition issue again (although in a much more subtle way).

`x` is also the name of a `typed` argument of the template this procedure is defined inside. Below, the template is invoked with non-top-level globals passed to `x`.

This leads to the `var substitutedGlobalSymbolGoesHere` to be treated as a procedure-level global. If those have a destructor (`legacyInit`, `instGenericInit`, etc. do have one), they're registered in the global destructor list, which in this case caused the globals to be destroyed twice (a double free that resulted in a crash).

**Pull request merged**